File: thisweek_masuipeo/q711/tyama_codeiq711_iter.kt
//usr/bin/env kscript $0 $@;exit
fun <T>reverse(a: Array<T>,start: Int,size: Int){
	val end=start+size-1
	for(i in 0..size/2-1){
		val z=a[start+i]
		a[start+i]=a[end-i]
		a[end-i]=z
	}
}
fun <T: Comparable<T>>unique_permutation(a0: Array<T>,n: Int):Sequence<Array<T>>{
	return sequence seq@{
		if(n<0||a0.size<n)return@seq
		var a=a0.copyOf()
		while(true){
			yield(a.sliceArray(0..n-1))
			reverse(a,n,a.size-n)
			var k = -1
			for(i in (a.size-2) downTo 0 step 1)if(a[i]<a[i+1]){k=i;break}
			if(k<0){
				//reverse(a,0,a.size)
				break
			}
			var l = -1
			for(i in a.size-1 downTo k+1 step 1)if(a[k]<a[i]){l=i;break}
			val z=a[k];a[k]=a[l];a[l]=z
			reverse(a,k+1,a.size-(k+1))
		}
	}
}
fun <T: Comparable<T>>unique_permutation(a: Array<T>):Sequence<Array<T>>{
	return unique_permutation(a,a.size)
}
fun main(args: Array<String>){
	val N=6
	var r=0
	val e0=Array<Int>(N*2){0}
	val f0=Array<Int>(N*2){0}
	for(i in 0..N-1){e0[N+i]=1;f0[N+i]=1}
	val e=IntArray(N*2+1)
	val f=IntArray(N*2+1)
	for(e_ in unique_permutation(e0)){
		for(i in 0..N*2-1)e[i+1]=e[i]+e_[i]
		for(f_ in unique_permutation(f0)){
			var flg=0
			var zero1=0
			var zero2=N
			var one1=0
			var one2=N
			for(i in 0..N*2-1){
				if(e_[i]==0)zero1+=1
				if(e_[i]==1)one1+=1
				if(f_[N*2-1-i]==0)zero2-=1
				if(f_[N*2-1-i]==1)one2-=1
				if(zero1==zero2)flg+=1
				if(one1==one2)flg+=1
			}
			if(flg>=2)r+=1
		}
	}
	println(r);
}

File: thisweek_masuipeo/q711/tyama_codeiq711_iter.nim
#[
/usr/bin/env nim c --nimcache:$HOME/.nimcache --opt:speed --run $0 $@;exit
#]#

import algorithm
import sequtils

proc unique_permutation[T](a0:seq[T],n0:ref int=nil):iterator:seq[T] =
 return iterator:seq[T] =
  var n=len(a0)
  if n0!=nil:
   n=n0[]
  if n<0 or len(a0)<n: return
  var a=a0.sorted
  while true:
   yield a[0..n-1]
   a=concat(a[0..n-1],a[n..len(a)-1].reversed)
   var k = -1
   block blk:
    for i in countdown(len(a)-2,0):
     if a[i]<a[i+1]:
      k=i
      break blk
    #a.reverse
    return
   for l in countdown(len(a)-1,k):
    if a[k]<a[l]:
     let z=a[k]
     a[k]=a[l]
     a[l]=z
     break
   a=concat(a[0..k],a[k+1..len(a)-1].reversed)

let N=6
var e0=concat(repeat(0,N),repeat(1,N))
var f0=concat(repeat(0,N),repeat(1,N))
#各Pは経路を表す。1が縦、0が横を表す。
var r=0
var e=repeat(0,N*2+1)
var f=repeat(0,N*2+1)
for e1 in unique_permutation(e0):
 for f1 in unique_permutation(f0):
  var flg=0
  var zero1=0
  var zero2=N
  var one1=0
  var one2=N
  for i in countup(0,N*2-1):
   zero1 += (if e1[i]==0: 1 else: 0)
   one1  += (if e1[i]==1: 1 else: 0)
   zero2 -= (if f1[N*2-1-i]==0: 1 else: 0)
   one2  -= (if f1[N*2-1-i]==1: 1 else: 0)
   if zero1==zero2: flg+=1
   if one1==one2: flg+=1
  if flg>=2: r+=1
writeLine(stdout,r) # 527552
